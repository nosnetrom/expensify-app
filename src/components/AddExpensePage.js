import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startAddExpense(expense);
    this.props.history.push('/'); // Redirect to home
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Add Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm 
            onSubmit = {this.onSubmit}
          />
        </div>
      </div>
    );
  };
};

/* Converted to a class, for testing purposes */
// const AddExpensePage = (props) => (
//   <div>
//     <h1>Add Expense</h1>
//     <ExpenseForm 
//       onSubmit = {(expense) => {
//         props.onSubmit(expense);
//         props.history.push('/'); // Redirect to home
//       }}
//     />
//   </div>
// );

const mapDispatchToProps = (dispatch) => ({ 
  startAddExpense: (expense) => dispatch(startAddExpense(expense)) 
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);