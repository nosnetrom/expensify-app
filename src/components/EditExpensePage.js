import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    // Dispatch the action
    this.props.editExpense(this.props.expense.id, expense);
    // Redirect to dashboard
    this.props.history.push('/');
  };
  onRemove = (event) => {
    this.props.removeExpense({id: this.props.expense.id});
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
      <ExpenseForm 
        expense={this.props.expense}
        onSubmit={this.onSubmit}
      />
      <button
        className="ExpenseItem__btnRemove"
        onClick={this.onRemove}
      >
        Remove expense
      </button>
    </div>
    );
  };
};

/*const EditExpensePage = (props) => {
  return (
    <div>
      <ExpenseForm 
        expense={props.expense}
        onSubmit={(expense) => {
          // Dispatch the action
          props.dispatch(editExpense(props.expense.id, expense));
          // Redirect to dashboard
          props.history.push('/');
          console.log('updated', expense);
        }}
      />
      <button
        className="ExpenseItem__btnRemove"
        onClick={(event) => {
          props.dispatch(removeExpense({id: props.expense.id}));
          props.history.push('/');
        }}
      >
        Remove expense
      </button>
    </div>
  )
};*/

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  };
};
const mapDispatchToProps = (dispatch, props) => ({ 
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  removeExpense: (data) => dispatch(removeExpense(data))
});


export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);