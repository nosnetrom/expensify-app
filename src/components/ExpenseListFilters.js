import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';

export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };
  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }));
  };
  onTextChange = (event) => {
    this.props.setTextFilter(event.target.value); // like store.dispatch()
  };
  onSortChange = (event) => {
    event.target.value === 'amount' ?
      this.props.sortByAmount() :
      this.props.sortByDate();
  };
  render() {
    return (
      <div>
        <input type="text" 
          className="ExpenseList__filters"
          value={this.props.filters.text}
          onChange={this.onTextChange} />
        <label htmlFor="selectSortBy"> Sort by: </label>
        <select 
          name="selectSortBy"
          value={this.props.filters.sortBy}
          onChange={this.onSortChange}>
            <option 
              value="date"
            >Date</option>
            <option 
              value="amount"
            >Amount</option>
        </select>
        <DateRangePicker 
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          onFocusChange={this.onFocusChange}
          focusedInput={this.state.calendarFocused}
          showClearDates={true}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
      </div>
    )
  };
};

const mapStateToProps = (state) => {
  return {
    filters: state.filters
  }
};
const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);