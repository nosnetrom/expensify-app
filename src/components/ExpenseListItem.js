import React from 'react';
import { Link } from 'react-router-dom';

const ExpenseListItem = ({id, description, amount, createdAt}) => (
  <p>
    <span className="ExpenseItem__description">
    <Link to={`/edit/${id}`}>{description}</Link>
    </span>
    <span className="ExpenseItem__amount">{amount}</span>
    <span className="ExpenseItem__createdAt">{createdAt}</span>
  </p>
)

export default ExpenseListItem;