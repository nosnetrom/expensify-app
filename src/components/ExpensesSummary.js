import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = ({expenseCount, expensesTotal}) => {
  const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
  return (
    <div className="page-header">
      <p className="page-header__title">
        Viewing <span>{expenseCount}</span> {expenseWord} totalling <span>{numeral(expensesTotal / 100).format('$0,0.00')}</span>.
      </p>
      <div className="page-header__actions">
        <Link className="btn-blue" to="/create">Add Expense</Link>
      </div>
    </div>
  );
};

const mapStateToProps = (state) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);

  return {
    expenseCount: visibleExpenses.length,
    expensesTotal: selectExpensesTotal(visibleExpenses)
  }
};

export default connect(mapStateToProps)(ExpensesSummary);
