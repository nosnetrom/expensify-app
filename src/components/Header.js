import React from 'react';
import { NavLink } from 'react-router-dom';

const Header = () => (
  <header>
    <h1>Expensify</h1>
    <p>
      <NavLink to="/" exact={true} activeClassName="isActive">Dashboard</NavLink>&nbsp;
      <NavLink to="/create" activeClassName="isActive">Create Expense</NavLink>&nbsp;
      <NavLink to="/help" activeClassName="isActive">Help</NavLink>
    </p>
  </header>
);

export default Header;