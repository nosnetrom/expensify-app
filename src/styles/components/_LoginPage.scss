.LoginPage {
  display: flex; 
  height: 100vh;
  text-align: center;
  line-height: 2.75;
}

.LoginPage p {
  width: 240px;
  height: 120px;
  margin: auto;
}