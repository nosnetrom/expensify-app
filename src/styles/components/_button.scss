button {
  border: none;
  cursor: pointer;
}
button:disabled {
  cursor: default;
}
.btn-blue {
  background-color: $blue;
  color: $white;
  display: inline-block;
  line-height: 1;
  padding: $sSize $xsSize;
  text-decoration: none;
}
