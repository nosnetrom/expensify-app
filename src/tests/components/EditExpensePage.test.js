import React from 'react';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';
import { EditExpensePage } from '../../components/EditExpensePage';

let editExpense, removeExpense, history, wrapper;

beforeEach(() => { // A jest global function, to run before each test module
  editExpense = jest.fn();
  removeExpense = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(<EditExpensePage 
    editExpense={editExpense} 
    removeExpense={removeExpense} 
    history={history}
    expense={expenses[0]} />);
});

test('test to render EditExpensePage correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

// test cases:
// test to render editExpense, test to edit expense (w/spies), test to remove (w/spies)
test('test EditExpensePage to correctly edit', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0]);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(editExpense).toHaveBeenLastCalledWith(expenses[0].id, expenses[0]);
});

test('test to removeExpense', () => {
  wrapper.find('button').last().simulate('click');
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(removeExpense).toHaveBeenLastCalledWith({
    id: expenses[0].id
  })
})