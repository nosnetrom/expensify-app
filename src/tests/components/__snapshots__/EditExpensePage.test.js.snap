// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test to render EditExpensePage correctly 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 195,
        "createdAt": 0,
        "description": "Gum",
        "id": "a",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    className="ExpenseItem__btnRemove"
    onClick={[Function]}
  >
    Remove expense
  </button>
</div>
`;
